src/main/java/com/iota/iri/service/API.java

```diff
@@ -805,48 +807,32 @@ private AbstractResponse getBalancesStatement(final List<String> addrss, final L
         }
 
         final Map<Hash, Long> balances = new HashMap<>();
-        final int index;
-        Snapshot referenceSnapshot;
-        synchronized (instance.milestone.latestSnapshot.snapshotSyncObject) {
-            referenceSnapshot = new Snapshot(instance.milestone.latestSnapshot);
-        }
-        index = referenceSnapshot.index();
-        for(Hash hash: hashes) {
-            if (!TransactionViewModel.exists(instance.tangle, hash)) {
-                return ErrorResponse.create("Tip not found: " + hash.toString());
-            }
-            if (!instance.ledgerValidator.isTipConsistent(referenceSnapshot, hash)) {
-                return ErrorResponse.create("Tips are not consistent");
-            }
-        }
+        instance.milestone.latestSnapshot.rwlock.readLock().lock();
```

Comment:
should be moved up to before L803? 
`hashes = Collections.singletonList(instance.milestone.latestSolidSubtangleMilestone);`

as the milestone can change between L803 and 810. so the index & milestone hash may mismatch.